File: backend/services/file_service.py
"""文件服务 - 处理笔记文件的读取"""
import os
from pathlib import Path
from typing import Optional
from config import NOTES_DIR, ALLOWED_EXTENSIONS


def get_note_tree() -> list:
    """
    获取笔记目录树
    返回结构: [{ name, type, path, children? }]
    """
    return _scan_directory(NOTES_DIR)


def _scan_directory(directory: Path, relative_base: Path = None) -> list:
    """递归扫描目录"""
    if relative_base is None:
        relative_base = NOTES_DIR
    
    items = []
    
    if not directory.exists():
        return items
    
    # 获取目录内容并排序（文件夹优先，然后按名称排序）
    entries = sorted(
        directory.iterdir(),
        key=lambda x: (not x.is_dir(), x.name.lower())
    )
    
    for entry in entries:
        # 跳过隐藏文件和目录
        if entry.name.startswith('.'):
            continue
        
        relative_path = entry.relative_to(relative_base)
        
        if entry.is_dir():
            children = _scan_directory(entry, relative_base)
            # 只添加非空文件夹或包含md文件的文件夹
            if children:
                items.append({
                    "name": entry.name,
                    "type": "directory",
                    "path": str(relative_path).replace("\\", "/"),
                    "children": children
                })
        elif entry.suffix.lower() in ALLOWED_EXTENSIONS:
            items.append({
                "name": entry.stem,  # 不带扩展名的文件名
                "type": "file",
                "path": str(relative_path).replace("\\", "/"),
            })
    
    return items


def get_note_content(note_path: str) -> Optional[str]:
    """
    获取笔记内容
    :param note_path: 相对于 notes 目录的路径
    :return: 笔记内容或 None
    """
    # 安全检查：防止目录穿越攻击
    safe_path = _safe_path(note_path)
    if safe_path is None:
        return None
    
    if not safe_path.exists():
        return None
    
    if safe_path.suffix.lower() not in ALLOWED_EXTENSIONS:
        return None
    
    try:
        return safe_path.read_text(encoding='utf-8')
    except Exception:
        return None


def _safe_path(note_path: str) -> Optional[Path]:
    """
    验证并返回安全的文件路径
    防止目录穿越攻击
    """
    try:
        # 规范化路径
        normalized = Path(note_path).as_posix()
        # 移除开头的斜杠
        normalized = normalized.lstrip('/')
        
        # 构建完整路径
        full_path = (NOTES_DIR / normalized).resolve()
        
        # 确保路径在 NOTES_DIR 内
        if not str(full_path).startswith(str(NOTES_DIR.resolve())):
            return None
        
        return full_path
    except Exception:
        return None


def search_notes(keyword: str) -> list:
    """
    搜索笔记内容
    :param keyword: 搜索关键词
    :return: 匹配的笔记列表
    """
    results = []
    keyword_lower = keyword.lower()
    
    for md_file in NOTES_DIR.rglob("*.md"):
        try:
            content = md_file.read_text(encoding='utf-8')
            if keyword_lower in content.lower() or keyword_lower in md_file.stem.lower():
                relative_path = md_file.relative_to(NOTES_DIR)
                # 提取匹配的上下文
                context = _extract_context(content, keyword)
                results.append({
                    "name": md_file.stem,
                    "path": str(relative_path).replace("\\", "/"),
                    "context": context
                })
        except Exception:
            continue
    
    return results


def _extract_context(content: str, keyword: str, context_length: int = 100) -> str:
    """提取关键词周围的上下文"""
    keyword_lower = keyword.lower()
    content_lower = content.lower()
    
    pos = content_lower.find(keyword_lower)
    if pos == -1:
        # 如果内容中没找到，返回开头部分
        return content[:context_length] + "..." if len(content) > context_length else content
    
    start = max(0, pos - context_length // 2)
    end = min(len(content), pos + len(keyword) + context_length // 2)
    
    context = content[start:end]
    if start > 0:
        context = "..." + context
    if end < len(content):
        context = context + "..."
    
    return context

File: backend-go/services/file_service.go
package services

import (
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"

	"mdxy-backend/config"
)

// TreeNode 目录树节点
type TreeNode struct {
	Name     string      `json:"name"`
	Type     string      `json:"type"`
	Path     string      `json:"path"`
	Children []*TreeNode `json:"children,omitempty"`
}

// SearchResult 搜索结果
type SearchResult struct {
	Name    string `json:"name"`
	Path    string `json:"path"`
	Context string `json:"context"`
}

// GetNoteTree 获取笔记目录树
func GetNoteTree() []*TreeNode {
	return scanDirectory(config.NotesDir, config.NotesDir)
}

// scanDirectory 递归扫描目录
func scanDirectory(directory string, relativeBase string) []*TreeNode {
	items := []*TreeNode{}

	// 检查目录是否存在
	if _, err := os.Stat(directory); os.IsNotExist(err) {
		return items
	}

	// 读取目录内容
	entries, err := ioutil.ReadDir(directory)
	if err != nil {
		return items
	}

	// 对条目进行排序（文件夹优先，然后按名称排序）
	// 这里简化处理，实际可以更复杂

	for _, entry := range entries {
		// 跳过隐藏文件和目录
		if strings.HasPrefix(entry.Name(), ".") {
			continue
		}

		fullPath := filepath.Join(directory, entry.Name())
		relativePath, _ := filepath.Rel(relativeBase, fullPath)

		if entry.IsDir() {
			children := scanDirectory(fullPath, relativeBase)
			// 只添加非空文件夹或包含md文件的文件夹
			if len(children) > 0 {
				items = append(items, &TreeNode{
					Name:     entry.Name(),
					Type:     "directory",
					Path:     strings.ReplaceAll(relativePath, "\\", "/"),
					Children: children,
				})
			}
		} else {
			// 检查文件扩展名是否允许
			ext := strings.ToLower(filepath.Ext(entry.Name()))
			if config.AllowedExtensions[ext] {
				name := strings.TrimSuffix(entry.Name(), ext)
				items = append(items, &TreeNode{
					Name: name,
					Type: "file",
					Path: strings.ReplaceAll(relativePath, "\\", "/"),
				})
			}
		}
	}

	return items
}

// GetNoteContent 获取笔记内容
func GetNoteContent(notePath string) string {
	// 安全检查：防止目录穿越攻击
	safePath := SafePath(notePath)
	if safePath == "" {
		return ""
	}

	// 检查文件是否存在
	if _, err := os.Stat(safePath); os.IsNotExist(err) {
		return ""
	}

	// 检查文件扩展名是否允许
	ext := strings.ToLower(filepath.Ext(safePath))
	if !config.AllowedExtensions[ext] {
		return ""
	}

	// 读取文件内容
	content, err := ioutil.ReadFile(safePath)
	if err != nil {
		return ""
	}

	return string(content)
}

// SafePath 验证并返回安全的文件路径，防止目录穿越攻击
func SafePath(notePath string) string {
	// 规范化路径
	normalized := filepath.Clean(notePath)
	// 移除开头的斜杠
	normalized = strings.TrimPrefix(normalized, "/")

	// 构建完整路径
	fullPath := filepath.Join(config.NotesDir, normalized)

	// 确保路径在 NOTES_DIR 内
	rel, err := filepath.Rel(config.NotesDir, fullPath)
	if err != nil || strings.HasPrefix(rel, "..") {
		return ""
	}

	return fullPath
}

// SearchNotes 搜索笔记内容
func SearchNotes(keyword string) []SearchResult {
	results := []SearchResult{}
	keywordLower := strings.ToLower(keyword)

	// 遍历 NOTES_DIR 目录下的所有 .md 文件
	err := filepath.Walk(config.NotesDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}

		// 只处理文件
		if !info.IsDir() {
			// 检查文件扩展名
			ext := strings.ToLower(filepath.Ext(path))
			if config.AllowedExtensions[ext] {
				// 读取文件内容
				content, readErr := ioutil.ReadFile(path)
				if readErr != nil {
					return nil
				}

				contentStr := string(content)
				contentLower := strings.ToLower(contentStr)

				// 检查是否包含关键词
				if strings.Contains(contentLower, keywordLower) || strings.Contains(strings.ToLower(info.Name()), keywordLower) {
					relativePath, _ := filepath.Rel(config.NotesDir, path)
					// 提取匹配的上下文
					context := extractContext(contentStr, keyword)
					results = append(results, SearchResult{
						Name:    strings.TrimSuffix(info.Name(), ext),
						Path:    strings.ReplaceAll(relativePath, "\\", "/"),
						Context: context,
					})
				}
			}
		}

		return nil
	})

	if err != nil {
		return results
	}

	return results
}

// extractContext 提取关键词周围的上下文
func extractContext(content string, keyword string) string {
	contextLength := 100
	keywordLower := strings.ToLower(keyword)
	contentLower := strings.ToLower(content)

	pos := strings.Index(contentLower, keywordLower)
	if pos == -1 {
		// 如果内容中没找到，返回开头部分
		if len(content) > contextLength {
			return content[:contextLength] + "..."
		}
		return content
	}

	start := pos - contextLength/2
	if start < 0 {
		start = 0
	}

	end := pos + len(keyword) + contextLength/2
	if end > len(content) {
		end = len(content)
	}

	context := content[start:end]
	if start > 0 {
		context = "..." + context
	}
	if end < len(content) {
		context = context + "..."
	}

	return context
}
